iceberg-rust/src/table/manifest_list.rs

```diff
@@ -57,20 +61,301 @@ impl<'a, 'metadata, R: Read> ManifestListReader<'a, 'metadata, R> {
+        let bytes: Cursor<Vec<u8>> = Cursor::new(
+            object_store
+                .get(&strip_prefix(snapshot.manifest_list()).into())
+                .await?
+                .bytes()
+                .await?
+                .into(),
+        );
+        ManifestListReader::new(bytes, table_metadata).map_err(Into::into)
+    }
+}
+
+/// A helper to write entries into manifest lists and files
+pub struct ManifestListWriter<'metadata> {
     table_metadata: &'metadata TableMetadata,
+    manifest_list_writer: apache_avro::Writer<'static, Vec<u8>>,
+    snapshot_id: i64,
     object_store: Arc<dyn ObjectStore>,
-) -> Result<impl Iterator<Item = Result<ManifestListEntry, Error>> + 'metadata, Error> {
-    let bytes: Cursor<Vec<u8>> = Cursor::new(
-        object_store
-            .get(&strip_prefix(snapshot.manifest_list()).into())
+    snapshot_uuid: String,
+    manifest_index: usize,
+    branch: Option<String>,
+    avro_schema: AvroSchema,
+}
+
+impl<'schema, 'metadata> ManifestListWriter<'metadata> {
+    /// New manifest list writer
+    pub fn try_new(
+        object_store: Arc<dyn ObjectStore>,
+        table_metadata: &'metadata TableMetadata,
+        branch: Option<String>,
+    ) -> Result<Self, Error> {
+        let manifest_list_schema = match table_metadata.format_version {
+            FormatVersion::V1 => manifest_list_schema_v1(),
+            FormatVersion::V2 => manifest_list_schema_v2(),
+        };
+        let snapshot_id = generate_snapshot_id();
+        let snapshot_uuid = uuid::Uuid::new_v4().to_string();
+        let manifest_list_writer = apache_avro::Writer::new(manifest_list_schema, Vec::new());
+
+        let partition_fields = table_metadata.current_partition_fields(branch.as_deref())?;
+        let avro_schema = ManifestEntry::schema(
+            &partition_value_schema(&partition_fields)?,
+            &table_metadata.format_version,
+        )?;
+
+        Ok(Self {
+            table_metadata,
+            manifest_list_writer,
+            snapshot_id,
+            snapshot_uuid,
+            object_store,
+            manifest_index: 0,
+            branch,
+            avro_schema,
+        })
+    }
+
+    /// Append a serializable ManifestListEntry
+    pub fn append_ser(&mut self, entry: &ManifestListEntry) -> Result<usize, Error> {
+        self.manifest_list_writer
+            .append_ser(entry)
+            .map_err(Error::from)
+    }
+
+    /// Create new manifests for each group of provided manifest entries and append them to the manifest list
+    pub async fn append_manifest_entries_to_new_manifests(
+        &mut self,
+        grouped_manifest_entries: Vec<Vec<ManifestEntry>>,
+    ) -> Result<(), Error> {
+        for entry_group in grouped_manifest_entries {
+            let manifest_location = format!(
+                "{}/metadata/{}-m{}.avro",
+                self.table_metadata.location, self.snapshot_uuid, self.manifest_index
+            );
```

Comment:
I think the error is coming from here. The manifest_index should be different for every manifest_group. It's overwriting the previous files.

Reply:
Ah yes, looks like I was forgetting to update manifest_index. It should be easy to add an extra test to also catch this.